Multi-path symex-only checker [blocks: 3795]

peterschrammel opened the pull request:
Based on #3585, only review last 4 commits

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Use goto verifier for program-only and show-vcc in CBMC

peterschrammel committed:
Add properties_result_from_symex_target_equation

Updates the property infos with properties
created by goto-symex and sets the status
of already determined properies.

peterschrammel committed:
Use goto verifier for program-only and show-vcc in JBMC

peterschrammel commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -548,6 +551,19 @@ int cbmc_parse_optionst::doit()
+    }
+  }
+
   if(cmdline.isset("validate-goto-model"))
```
Validate should be run before symex.

tautschnig commented on src/goto-checker/multi_path_symex_only_checker.cpp:
```diff
@@ -0,0 +1,114 @@
+      get_memory_model(options, ns);
+    memory_model->set_message_handler(ui_message_handler);
+    (*memory_model)(equation);
+  }
```
I don't think this is of any use for this checker?

peterschrammel replied:
It won't have any impact on the property status, but I'd like still want to see the entire SSA.

tautschnig replied:
Yes, but doing unnecessary work should at least be explicitly documented via comments.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8253fa4b19b91f0d0cee110dfb1b18c0f3e27093).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97176111

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b197d357e7c3fbdd3f3ffb843578619f27b4e9bc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97179981

tautschnig commented:
The depended-on PRs have been merged, this needs a rebase and possibly further work to make CI pass.

peterschrammel commented:
~@tautschnig, regarding https://github.com/diffblue/cbmc/pull/3794/files#diff-e6155d797631f225220cb497ee60def3R232 and https://github.com/diffblue/cbmc/pull/3794/files#diff-e6155d797631f225220cb497ee60def3R260.
I'd like not to change any current behaviour during this refactoring. However, that requires some special-casing to adapt to some of the current behaviours. I can leave the comments in and change the current behaviour later or I can leave them out and create a separate PR now that changes the current behaviour.~

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 92f74e4b7409d38ebb974fe22354d5d398b4d581).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97233627

peterschrammel committed:
Do collect properties from inlined functions

They are not inlined using goto_partial_inline anymore.

peterschrammel committed:
Add multi path symex only checker

multi_path_symex_only_checkert is a
bounded model checking algorithm that determines the
status of properties through symbolic execution and
constant propagation.
I.e. it doesn't call the SAT solver.

If desired CBMC could expose this to support a symex-only
BMC algorithm.

For now, this is used to implement show-vcc and program-only
(which are rather orthogonal to the BMC algorithm used,
but currently assume symex-only).

peterschrammel committed:
Goto checkers for Java require further preprocessing

peterschrammel committed:
Allow UNKNOWN as neutral element in property status updates

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7bf4d1293faef1f9ade1a88ccd0912f04775e77f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97960133

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2061df5b5557972c81cf6ab658b283a7df39dc00).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97972656

smowton commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -225,3 +225,53 @@ void slice(
+    }
+    else
+    {
+      property_infot &property_info = properties.at(property_id);
```
Repeated lookup should probably be replaced with
```
auto emplace_result = properties.emplace(
  property_id, property_infot{step.source.pc, step.comment, status});
if(emplace_result.second)
{
  ...
}
else
{
  property_infot &property_info = emplace_result.first->second;
  ...
}
```

smowton commented on src/goto-checker/bmc_util.cpp:
```diff
@@ -225,3 +225,53 @@ void slice(
+
+  for(auto &property_pair : properties)
+  {
+    if(property_pair.second.status == property_statust::NOT_CHECKED)
```
Is there a chance the status was `PASS` before the loop above changed it to `NOT_CHECKED`, and therefore we might spuriously note it as changed? I *think* that can't happen because it's hard to imagine that `|=`result with either `UNKNOWN` or `PASS` on the RHS would turn `PASS` -> `NOT_CHECKED`, but just want to check I've understood.

peterschrammel replied:
No, cannot happen. `|=` is supposed to be monotonic, i.e. once you have a `PASS`, you cannot go back to something undetermined anymore.

smowton commented on src/goto-checker/bmc_util.h:
```diff
@@ -67,6 +69,13 @@ void slice(
   const optionst &,
   ui_message_handlert &);
 
+/// Sets property status to PASS or FAIL for properties whose
```
Does it? It appears to only set properties to `PASS`

smowton commented on jbmc/src/java_bytecode/java_multi_path_symex_only_checker.cpp:
```diff
@@ -0,0 +1,40 @@
+  : multi_path_symex_only_checkert(options, ui_message_handler, goto_model)
+{
+  // unwinds <clinit> loops to number of enum elements
+  if(options.get_bool_option("java-unwind-enum-static"))
```
Maybe factor this code with the identical code in jbmc?

peterschrammel replied:
Factored out into separate file. The instance of the code in jbmc_parse_options will disappear with the `do_language_agnostic_bmc` call.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 380945460556b05f37d05a80de2219c022f28096).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98183898
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.
